exonum/src/events/network.rs

```diff
@@ -369,47 +380,73 @@ impl Listener {
             let network_tx = network_tx.clone();
 
             let handshake = NoiseHandshake::responder(&handshake_params);
-            let stream = handshake.listen(sock).flatten_stream();
-
-            let connection_handler = stream
-                .into_future()
-                .and_then(Ok)
-                .map_err(|e| e.0)
-                .and_then(move |(raw, stream)| match raw.map(Any::from_raw) {
-                    Some(Ok(Any::Connect(msg))) => Ok((msg, stream)),
-                    Some(Ok(other)) => Err(other_error(&format!(
-                        "First message is not Connect, got={:?}",
-                        other
-                    ))),
-                    Some(Err(e)) => Err(into_other(e)),
-                    None => Err(other_error("Incoming socket closed")),
-                })
-                .and_then(move |(connect, stream)| {
-                    trace!("Received handshake message={:?}", connect);
-                    let event = NetworkEvent::PeerConnected(addr, connect);
-                    let stream = network_tx
-                        .clone()
-                        .send(event)
-                        .map_err(into_other)
-                        .and_then(move |_| Ok(stream))
-                        .flatten_stream();
-
-                    stream.for_each(move |raw| {
-                        let event = NetworkEvent::MessageReceived(addr, raw);
-                        network_tx.clone().send(event).map_err(into_other).map(drop)
-                    })
-                })
-                .map(|_| {
-                    // Ensure that holder lives until the stream ends.
-                    let _holder = holder;
+            let connection_handler = handshake
+                .listen(sock)
+                .and_then(move |(sock, key)| {
+                    let (_, stream) = sock.split();
+                    stream
+                        .into_future()
+                        .map_err(|e| e.0)
+                        .and_then(|(raw, stream)| (Self::parse_connect_msg(raw), Ok(stream)))
+                        .and_then(move |(connect, stream)| {
+                            trace!("Received handshake message={:?}", connect);
+
+                            Self::process_incoming_messages(
+                                stream,
+                                network_tx,
+                                connect,
+                                ConnectInfo {
```

Comment:
Very minor nit: Maybe, it could make sense to rename local vars `addr` and `key` to `address` and `public_key`, respectively, in order to use the shorthand constructor here?